### ISSUE TITLE
Chain independence: tracking issue (the framework's known open limit)

### ISSUE THREAD

**Issue by alizahidraja** [alizahidraja/isnad#54]
## Status: open tracking issue (do not close)

This is the framework's single most important *known* limitation, stated publicly so it is never mistaken for a solved problem.

## The open critique, in neutral terms

Corroboration (mutābaʿāt) upgrades a claim only when the corroborating chains are **independent**. The current `SharedLineageDetector` checks three structural signals — shared narrator identity, shared model family, and shared upstream source — and discounts corroboration when any is detected.

But two chains can share **none** of those three signals and still fail together:

1. **Correlated training data / model blind spots** — different models, different families, no shared source, yet both reproduce the same corpus-wide mistake.
2. **Content-level correlated error** — two genuinely independent sources repeating the same *received* error (the classical madār problem in a different guise).

The paper's own §7 says this plainly: "a genuinely correlated pair that shares none of those three attributes would still pass, and content-level correlated error — two independent sources repeating the same received mistake — is not addressed at all." This is a *structural* limit of topology-based independence, not an implementation bug.

## Related

- #1 (closed) — "Chain independence cannot be inferred from chain topology"
- #44 — "Share evidence, never grades — federation without a Sybil surface" (evidence exchange makes independence *verifiable* rather than *assumed*)
- #47 — provenance interop (SLSA/SBOM attestations can carry training-data provenance)
- `docs/case-study-xz-sleeper-narrator.md` §4 — the sock-puppet echo as the *detectable* case; the correlated-blind-spot case is the *undetectable* one.

## Candidate approaches (none chosen yet)

1. **Attestation-backed independence** — require each corroborating chain to carry a signed provenance attestation (SLSA/SBOM), and treat "independent" as "attested-distinct lineage," not "disjoint narrator IDs." (#47)
2. **Calibration, not detection** — stop claiming binary independence; emit a *corroboration-strength* score that decays with the estimated prior probability of shared lineage, and cap all corroboration at HASAN regardless (already partially true). This is an honesty change, not a detection mechanism.
3. **Content-level madār detection** — detect when multiple "independent" sources share an identical normalized error pattern (a fingerprint of a common upstream), and discount those. Hard: requires distinguishing "both correct" from "both wrong in the same way."

The point of this issue is the *disclosure*: independence is an assumption the framework makes, and no amount of topology checking fully discharges it. Anyone building on ISNAD should read this before relying on corroboration in a high-stakes setting.

**Comment by alizahidraja** [comment]
Slice implemented: #83 (merged) made the corroboration engine score unknown lineage below the gate — independence is now *demonstrated*, not assumed from empty metadata. Docs follow-up in #100. This issue stays open: the correlated-blind-spot and content-level-correlated-error cases remain undetectable from topology (paper §7).

**Comment by alizahidraja** [comment]
## Status update (audit pass, 2026-08-26)

One of the three candidate approaches is now chosen and tracked as a child issue:

- **#125 — "calibration, not detection"** is the selected near-term action: carry the
  continuous independence score (and its provenance) through `CorroborationResult`
  instead of collapsing to a boolean gate. It keeps the HASAN cap and the
  unknown-lineage-below-gate behaviour unchanged.

The other two candidates (attestation-backed via #47; content-level madār detection)
remain open here and are not yet scheduled. This issue stays open — the
correlated-blind-spot and content-level-correlated-error cases remain undetectable
from topology (§7).

**Comment by alizahidraja** [comment]
## Audit pass (2026-08-26): classical grounding strengthens the 'why isnād over a weighted DAG' answer

One addition to this tracking issue's candidate list, from the multi-persona audit of the
independence strategy: the **common-link (madār) mapping is real scholarship, not decoration**.

- Common-link analysis (the Schacht–Juynboll line) is precisely about detecting
  apparently-distinct chains that converge on a single origin.
- Tawātur requires genuinely separate chains as a *hard* criterion, not a soft one.
- A shared base model is a common link.

This is the strongest available answer to "what does isnād buy over a weighted DAG" —
reputation systems collapse correlation into one score and ignore the convergence
failure mode the tradition spent centuries on. It should be cited carefully (as the
scholars' *method*, not as an endorsement of any specific revisionist thesis), and
framed as one leg alongside the two-axis split (#124) and the decision matrix.

The concrete implementation now lives in #125 (upgraded: disjointness discount →
N_eff, emit-to-trace, unify the callback's 4-signal check into the core detector).

**Comment by alizahidraja** [comment]
## Resolution — the limit is now a parameter + a narrow detectable case (2.10.0)

The classical scholars never proved two narrators were independent. Tawātur is defined as
"a number of reporters so large that collusion on a lie is inconceivable" — a statement
about *joint failure probability*, not per-pair independence. We translated exactly that.

**Two changes shipped (PR #176):**

### 1. The tawātur discount (N_eff)

`CappedCorroborationPolicy` now takes `shared_blind_spot_prior` (default **0.20**): the prior
probability that a nominally-independent chain shares an *unobservable* correlated failure
(shared training data, a shared blind spot). Every corroborating chain's witness weight is
scaled by `(1 − prior)`, so **no chain ever earns full witness credit** — the honest tawātur
position that corroboration requires *more* chains than the nominal count suggests.

- `shared_blind_spot_prior=0.0` reproduces the pre-#54 behavior exactly (backward compatible).
- `shared_blind_spot_prior=1.0` disables corroboration entirely.
- `CorroborationResult` now reports `effective_witnesses` and `shared_blind_spot_prior`, so a
  caller sees "2 nominal chains → 1.6 effective witnesses (prior 0.20)", not a bare boolean.

This converts the two *unobservable* failure classes (correlated training data, shared blind
spots) from "undetectable" to "priced in as a stated prior." Operators with attestation (#47)
or evidence exchange (#44) may lower the prior; it must not be silently assumed away.

### 2. Content-level madār (the detectable half)

`core/content_madar.py` handles the case that *is* observable: when a claim is CONTRADICTION
against the corpus, and a "corroborating" chain repeats the **same specific error** (identical
wrong number, or identical dropped/flipped negation), that is a fingerprint of a common
upstream — a received mistake echoed through distinct routes, not independent confirmation.

- Correct agreement is **not** flagged (that's what independent corroboration looks like).
- Unverifiable claims are **not** flagged (content-madar is undetectable for novel claims, and
  that limit is stated, not papered over).

### What this means for the paper (§7)

Instead of "a genuinely correlated pair that shares none of those three attributes would still
pass, and content-level correlated error is not addressed at all," v2 can now say:

> Independence is priced, not assumed. Corroboration is discounted by an explicit
> shared-blind-spot prior (default 0.20), grounded in the classical tawātur criterion that
> corroboration must outnumber plausible collusion, and content-level correlated error is
> detected where it is checkable (a shared error fingerprint against the corpus). For novel,
> corpus-uncheckable claims, the prior remains the only guard — stated, not hidden.

This issue stays open as the tracking/disclosure issue, but it is no longer "none chosen yet" —
the tawātur discount is shipped and the madār detection is shipped.

**Comment by alizahidraja** [comment]
## Update 2 (2.10.1): the tawātur discount is now witness-type-aware

Deepened the #54 resolution with the shāhid/mutābaʿa distinction:

- **mutābaʿa** (a narrator confirming from the same teacher) = weak corroboration → high blind-spot prior.
- **shāhid** (a different companion confirming the same meaning) = strong corroboration → low prior.

`CappedCorroborationPolicy.BLIND_SPOT_MATRIX` keys the prior by narrator type: same-kind
(model+model 0.25) charges more than cross-kind (model+human 0.05, model+source 0.08). The
engine derives each chain's dominant narrator type from metadata and threads per-chain priors
through the decision, so `effective_witnesses`/`effective_weight` agree with the actual discount.
Every cell is a stated default to calibrate, never a truth.

**Honesty note (recorded, not hidden):** `detect_content_madar` (shipped 2.10.0) still has no
call site — its wiring into the API's critic verdicts is the next, separate follow-up. The
function is tested; it is not yet operational end-to-end.

**Comment by alizahidraja** [comment]
## STATUS UPDATE — 29 Aug 2026 (substantial progress; still not "solved")

The open critique below stands, but the *addressable* parts have now shipped
(v2.8.0 → v2.11.0). This issue remains open because the *undetectable* half is
structural, not because nothing was done.

**Now shipped:**

1. **Attested lineage** (#54, PR #83, v2.6.1) — absent `model_family` /
   `upstream_source` is now scored *unknown* (below the gate), so unattested
   chains no longer silently corroborate.
2. **Document-hash madār detection** (#125, v2.8.0) — two chains that retrieved
   the same document are one source, not two.
3. **Disjointness discount** (#125, v2.8.0) — corroboration weighted by an
   independence score.
4. **Tawātur discount (N_eff)** (#54, v2.10.0) — `shared_blind_spot_prior`
   (default 0.20) prices in the unobservable shared-failure probability; every
   chain's witness weight is scaled by (1 − prior); `effective_witnesses` reported.
5. **Witness-type-aware priors** (#54, v2.10.1) — shāhid vs mutābaʿa: pairwise
   blind-spot priors keyed by narrator type (model+model 0.25, model+human 0.05,
   model+source 0.08).
6. **Content-level madār detection** (`core/content_madar.py`, v2.10.0) — the
   *detectable* half: when the base claim is CONTRADICTION and a corroborator
   repeats the *same error* (identical wrong number / flipped negation), it is a
   fingerprint of a common upstream, not independent confirmation.

**Still open (honest):**

- Correlated training data / model blind spots across distinct families — the
  *undetectable* half. `detect_content_madar` is exported and tested but **not
  yet wired into `CorroborationEngine`'s decision path** (the shādhdh gate covers
  the base-claim contradiction case; the "corroborator repeats the same error"
  fingerprint is not threaded through yet). Tracked in #51 for the v2 paper.
- #44 (federation / evidence exchange) and #47 (SLSA/SBOM provenance interop)
  remain the attestation-backed path toward making independence *verifiable*
  rather than *assumed*.

**Comment by alizahidraja** [comment]
UPDATE 2026-08-29 (v2.12.0): the "detect_content_madar exported but not wired" gap is now closed — `CorroborationEngine` withholds corroboration when a corroborating chain repeats the base claim's same error, and reports `content_madar_detected`. The undetectable half (correlated training data across distinct model families) remains the open limit.

**Comment by alizahidraja** [comment]
Panel re-audit (2026-08-29, market+scope+vision lanes): status unchanged — correctly deferred until real users exist. See issue #130 for the ranked rationale. Key finding: #54's undetectable half (correlated training data across distinct model families) is active 2026 research (arXiv 2604.07650 'behavioral entanglement') and nobody productizes it — the most on-moat of the five, but research-stage with no buyer yet.

**Comment by alizahidraja** [comment]
## Deep research sweep — chain independence (2026-08-30)

10-person panel (rijāl scholar · ML-diversity · LLM-safety · supply-chain · data-provenance · Sybil · dedup · Bayesian · architect · red-team), each grounded in the repo and the findings below. **This is a planning/decision write-up — no code changed.**

### What the research says (what works, what doesn't)

| # | Finding | Source | Consequence for ISNAD |
| --- | --- | --- | --- |
| 1 | When several LLMs are wrong on the same prompt, they often pick the **same wrong answer** (~60% joint-error agreement). Vendor/architecture diversity is **not** independence — strong cross-family correlation (Claude×Gemini, GPT-4o×Claude), and *more* accurate models correlate *more*. | arXiv 2506.07962 | "Attested-distinct lineage" ≠ independent error; a flat prior under-prices the best models. |
| 2 | To measure: **double-fault rate** (P *all* fail together), **Q-statistic** (Q>0 = positively correlated), agreement-on-errors. Pairwise correlation **misses multi-model co-failure**; the key quantity is P(all fail together). | ICML 2025 poster 44225 | The current per-witness `(1−prior)` product under-counts common-cause failure. |
| 3 | SLSA + SBOM are complementary (a "dataset SBOM" + a provenance attestation referencing it by digest) — but SLSA is a **software**-supply-chain tool, an awkward adaptation for training data. | slsa.dev | Attestation proves build lineage, not training-data membership. |
| 4 | Sybil resistance requires tying influence to a **scarce resource** (compute/stake/trusted admission), never raw count. LLM-agent attacks: identity inflation, collusive corroboration, committee capture, reputation laundering. | ReCon et al. | Corroboration-by-count is cheap to fake; admission must be scarce. |
| 5 | Ensemble diversity metrics (disagreement, Q, double-fault) are **entangled with accuracy**, not pure independence signals; newer "focal diversity" (LLM-TOPLA) is emerging. | JMLR 23-0041 | Per-claim detection is impossible; only a population-level prior is honest. |
| 6 | Watermarking proves model **provenance** (was this text likely from model X), **not** training-data membership. | Stanford CRFM | No available signal proves a model did *not* train on a corpus. |

### 10-person debate — verdict per candidate

| Candidate | Verdict (count) | The refinement that survived |
| --- | --- | --- |
| **A — Attestation-backed independence** | **REFINE** (10/10) | Attestation is a **one-sided falsifier + admission gate**, never "independent." SBOM *content-overlap* → hard madār (score 0.0); attested-distinct lineage *lowers* `shared_blind_spot_prior` but never to zero; issuers must be trusted/scarce. Relabel score 1.0 from "demonstrably independent" to "no shared *declared* lineage detected." |
| **B — Calibration, not detection** | **SUPPORT** (10/10) — **the spine** | (1) ordinal output only — strip `effective_weight`/`effective_witnesses` floats from `reason` (they brush the no-numeric-confidence moat); (2) replace the flat `0.20` prior with a **measured** P(all-fail-together)/double-fault/Q per witness-type pair, tagged `assumption`/prior; (3) **joint** failure discount, not pairwise; (4) prior scales with narrator accuracy; (5) keep the HASAN cap. |
| **C — Content-level madār** | **SUPPORT as falsifier, REFINE recall** | Keep the `CONTRADICTION` gate (error-identity ≠ answer-identity — correct). Widen the fingerprint beyond numbers+negation to **error class** (wrong entity / date / relation / fabricated citation) + paraphrase. Rename `content_madar_detected` → `shared_error_detected` (evidence *consistent with* madār, not proof of a common upstream). Feed its measured rate into B's prior. |
| **New D1 — Empirical co-failure prior** | — | Measure double-fault/Q on a held-out labeled slice per narrator pair; feed B's prior; tag `OBSERVED`, not `PRIOR`. |
| **New D2 — Scarce-resource weighting** | — | Corroboration credit scales with costly admission (signed provenance / stake / trusted admission), never raw corroborator count. |
| **New D3 — Non-model witness** | — | Model-only corroboration never clears the independence gate; require one non-model witness (human / verified instrument with attested non-derived corpus). |

### Decision

**B is the governing mechanism; A and C become inputs that *lower* B's prior or *withhold* a verifiable echo; corroboration stays downgrade-only and ordinal.** Concretely, the future implementation plan (not now):

1. **Recalibrate the prior** — replace the hand-set `BLIND_SPOT_MATRIX` (0.08/0.20/0.25) with operator-measured double-fault/Q per witness-type pair, floored above zero, surfaced as an ordinal tier (`LOW/MOD/HIGH/UNKNOWN`) tagged as an assumption.
2. **Joint-failure discount** — model `P(all corroborators fail together) = (1−p)·∏errᵢ + p·err_base`, not the per-witness `score·(1−prior)·log(err)` product.
3. **Attestation as falsifier** — SBOM digest-overlap joins the madār check; attested-distinct lineage lowers (never zeroes) the prior; empty/missing SBOM stays `UNKNOWN` (0.5).
4. **Widen content-madār** to error-class + paraphrase, kept oracle-gated, renamed `shared_error_detected`.
5. **Moat fixes** (should land with any of the above): strip float weights from `reason`; relabel the 1.0 independence score.

### What would disprove this

A labeled benchmark showing `P(all fail together | attested-distinct, independently-admitted) ≈ 0` across many trials would justify binary independence (A as-is); conversely, a calibration set where the measured prior does **not** transfer to held-out novel claims would mean the calibration is fake precision — fall back to the flat prior.

**Comment by alizahidraja** [comment]
## Joint-failure discount — 6-persona panel outcome (2026-08-30)

Planning/decision only — no code changed. Panel: bayesian · rijal-scholar · architect · ml-diversity · red-team · honesty.

### Converged formula

For the admitted corroborator set (score >= 0.8, grade != MAWDU, min-gate HASAN):

```
e_i    = ERROR_PROBS[grade_i]          # SAHIH .01 / HASAN .10 / DAIF .30
score_i = independence score           # kept as the exponent (per-chain discount)
p       = max_i blind_spot_prior_for(base_type, corr_type_i)   # set-level shared-blind-spot prior
err_base = SHARED_MODE_ERROR = 0.10    # the shared mode fails at the HASAN rate (fixed constant)

P_all           = (1 - p) * e(base) * PROD(e_i ** score_i)  +  p * err_base
effective_weight = ln(P_all) / ln(ERROR_PROBS[HASAN])
fire iff effective_weight >= MIN_EFFECTIVE_WEIGHT (2.0)      # <=> P_all <= 0.01
```

At `p=0` this reduces to today's `e(base) * PROD(e_i ** score_i)` (backward compatible). One helper called by both `compute_corroborated_grade` and `_compute_effective_weight`, deleting the duplicated pairwise sum. `effective_witnesses` stays pairwise (descriptive count).

### BLOCKER (P0 — must resolve before coding)

With the **default** flat prior `p=0.20`, a DAIF base + 2 independent HASAN corroborators gives:

```
P_all = 0.80 * 0.30 * (0.10 * 0.10) + 0.20 * 0.10 = 0.0024 + 0.02 = 0.0224
w     = ln(0.0224) / ln(0.10) = 1.65  <  2.0   -> NO upgrade
```

So the joint model makes corroboration **dead at the default prior** — it only fires for attested cross-kind independence (p ≈ 0.05, w ≈ 2.11). The panel was unanimous this is a breaking tightening ("not point-release-safe") and recommended a **calibration sweep before coding**.

### The calibration decision (open)

The joint-failure discount cannot be calibrated without a **measured co-failure prior** — the double-fault rate / Q-statistic per witness-type pair measured on a held-out labeled slice. That measurement is what turns the flat 0.20 guess into a grounded per-pair prior. So the two remaining #54 items are coupled:

1. **measured co-failure prior** (build the harness, measure P(all fail together) per pair, store as an operator-local calibration record — the `OBSERVED` provenance), then
2. **joint-failure discount** (use that measured prior as `p`).

Until (1) exists, (2) is an honest-to-goodness research task, not a point-release.

### P1 (pre-existing, separate) — float leak

`api/endpoints/claims.py` serializes `effective_weight`, `effective_witnesses`, `shared_blind_spot_prior`, and per-chain `independence` floats into `corroboration_result`, returned verbatim by `GET /claims/{claim_id}`. This brushes the "no numeric confidence" moat. Recommended: redact these floats from the public record (keep them internal), or gate them behind an audit view. Tracked as its own issue.
